### PR TITLE
feat: experimental SSO support for Google, Microsoft, and OIDC

### PR DESCRIPTION
**Pull request by cheahjs**
# Pull Request Checklist

**Before submitting, make sure you've checked the following:**

- [x] **Target branch:** Please verify that the pull request targets the `dev` branch.
- [ ] **Description:** Provide a concise description of the changes made in this pull request.
- [ ] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [ ] **Documentation:** Have you updated relevant documentation [Open WebUI Docs](https://github.com/open-webui/docs), or other documentation sources?
- [ ] **Dependencies:** Are there any new dependencies? Have you updated the dependency versions in the documentation?
- [x] **Testing:** Have you written and run sufficient tests for validating the changes?
- [ ] **Code review:** Have you performed a self-review of your code, addressing any coding standard issues and ensuring adherence to the project's coding standards?
- [x] **Label:** To cleary categorize this pull request, assign a relevant label to the pull request title, using one of the following:
  - **BREAKING CHANGE**: Significant changes that may affect compatibility
  - **build**: Changes that affect the build system or external dependencies
  - **ci**: Changes to our continuous integration processes or workflows
  - **chore**: Refactor, cleanup, or other non-functional code changes
  - **docs**: Documentation update or addition
  - **feat**: Introduces a new feature or enhancement to the codebase
  - **fix**: Bug fix or error correction
  - **i18n**: Internationalization or localization changes
  - **perf**: Performance improvement
  - **refactor**: Code restructuring for better maintainability, readability, or scalability
  - **style**: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc.)
  - **test**: Adding missing tests or correcting existing tests
  - **WIP**: Work in progress, a temporary label for incomplete or ongoing work

# Changelog Entry

### Description

Addresses #483

Adds support for SSO logins with Google, Microsoft, or an OIDC provider.

Env vars:

* `ENABLE_OAUTH_SIGNUP` - allow accounts to be created when logging in with oauth
* `OAUTH_MERGE_ACCOUNTS_BY_EMAIL` - allow logging in to the account keyed by the email address present in the id token. This is less secure because handling the provided email address requires a lot of care that I have not put in.
* Google
  * see: https://support.google.com/cloud/answer/6158849?hl=en
  * GOOGLE_CLIENT_ID - Google OAuth client ID
  * GOOGLE_CLIENT_SECRET - Google OAuth client secret
  * Set redirect URI to `<backend>/oauth/google/callback`
* Microsoft
  * see: https://learn.microsoft.com/en-us/entra/identity-platform/quickstart-register-app
  * Only supports either Entra or personal apps, but not combined apps
  * MICROSOFT_CLIENT_ID - Microsoft OAuth client ID
  * MICROSOFT_CLIENT_SECRET - Microsoft OAuth client secret
  * MICROSOFT_CLIENT_TENANT_ID - Microsoft tenant ID, use `9188040d-6c67-4c5b-b112-36a304b66dad` for personal accounts
  * * Set redirect URI to `<backend>/oauth/microsoft/callback`
* OIDC
  * OAUTH_CLIENT_ID - OIDC client ID
  * OAUTH_CLIENT_SECRET - OIDC client secret
  * OPENID_PROVIDER_URL - OIDC well known URL, eg https://accounts.google.com/.well-known/openid-configuration 
  * OAUTH_PROVIDER_NAME - Name of the provider to show on the UI, defaults to SSO
  * * Set redirect URI to `<backend>/oauth/oidc/callback`

As email addresses are not generally trusted - authentication is matched against the `sub` field of the ID token unless `OAUTH_MERGE_ACCOUNTS_BY_EMAIL` is enabled.

### Added

- OIDC support - Admins can now configure login via Google, Microsoft, or a custom OIDC provider.

### Screenshots or Videos

image
image

Example deployment: https://jscheah-open-webui.hf.space/

### CODE REVIEW

**Comment by Yanyutin753** [comment]
That looks awesome, so try supporting github logins🥳🥳🥳

**Comment by kroonen** [comment]
SHIP SHIP SHIP!

**Comment by radderz** [comment]
It would be great if SSO can be made the default so that users never see the login screen, i.e. they are redirected to SSO by default. I.e. Entra ID for a company. And if the user doesn't already exist to create them.

**Comment by mjtechguy** [comment]
Would love to see this. This would def take OWUI to the next level!

**Comment by radderz** [comment]
> It would be great if SSO can be made the default so that users never see the login screen, i.e. they are redirected to SSO by default. I.e. Entra ID for a company. And if the user doesn't already exist to create them.

For this default style setting, admins would want a way to login without the SSO, like via a special URL that allows an admin to administer the system without the SSO in case something has gone wrong with SSO. Either that or way to disable default SSO through config/environment variables.

File: backend/apps/webui/models/users.py [hunk 1]
```diff
@@ -114,9 +120,23 @@ def get_user_by_api_key(self, api_key: str) -> Optional[UserModel]:
...
+    def get_user_by_email(
+        self, email: str, oauth_user: bool = False
+    ) -> Optional[UserModel]:
+        try:
+            conditions = (
+                (User.email == email, User.oauth_sub.is_null())
+                if not oauth_user
+                else (User.email == email)
+            )
+            user = User.get(conditions)
```
Comment (franciscovelez) [inline backend/apps/webui/models/users.py:132]: This condition is wrong. You are passing the tuple directly to the query which is causing a `peewee.OperationalError`. I suppose it should work prepending the `*` :
```python
user = User.get(conditions)
```

**Comment by tjbck** [comment]
THIS IS AMAZING 🤩 Much needed addition to Open WebUI, Thanks a bunch!

**Comment by justinh-rahb** [comment]
Man, this guy ***OAuths***

**Comment by leiguangyan** [comment]
@cheahjs  HI, Can you help provide a Docker-Compose deployment example? I've been testing and deploying but I'm getting this error again. I'm guessing this may be due to my docker-compose file being written incorrectly. and it keeps getting prompted that your provider did not provide a trusted header, please contact your administrator for help.

**Comment by justinh-rahb** [comment]
https://docs.openwebui.com/tutorial/sso#oauth

**Comment by leiguangyan** [comment]
@justinh-rahb  ok,thanks

**Comment by leiguangyan** [comment]
@justinh-rahb  HI，I still get the same error, my keycloak is in default configuration and has not been modified. Will there be a tutorial to support private deployment of keycloak authentication in the next version?

**Comment by justinh-rahb** [comment]
PR welcome if you figure it out and write one

**Comment by devdev999** [comment]
> @justinh-rahb HI，I still get the same error, my keycloak is in default configuration and has not been modified. Will there be a tutorial to support private deployment of keycloak authentication in the next version?

Did you create a client in keycloak? And pass the client secret to webui?

**Comment by leiguangyan** [comment]
Yes, I created it, but what seems to be returned is the key. Email not filtered out

**Comment by skeretna** [comment]
Using integration with Microsoft, can access be given or denied to an entire AD group? how can this be implemented?

**Comment by oliverbob** [comment]
Is this already a part of the the latest oi? I didn't see it in my last build. How do I enable it?

**Comment by Classic298** [comment]
@oliverbob check the docs

**Pull request merged**